File: app/src/main/java/com/kent/university/privelt/utils/sensors/SensorHelper.java
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.kent.university.privelt.utils.sensors;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.kent.university.privelt.model.Application;
import com.kent.university.privelt.model.Sensor;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class SensorHelper {

    private static boolean isSystemPackage(ApplicationInfo applicationInfo) {
        return ((applicationInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0);
    }

    private static String parseName(String name) {
        if (name == null || name.isEmpty())
            return name;
        String appName = name.substring(name.lastIndexOf(".") + 1).replaceAll("Application|App", "");
        if (appName.isEmpty())
            return name;
        else
            return appName;
    }

    private static boolean checkIfApplicationHasPermission(Application application, String sensor) {
        for (String permission : application.getPermissions()) {
            if (permission.toLowerCase().contains(sensor.toLowerCase()))
                return true;
        }
        return false;
    }

    public static List<Sensor> getSensorsInformation(Context context) {
        List<Application> applications = getApplicationsInformation(context);
        List<Sensor> sensors = Arrays.asList(Sensor.values());

        for (Sensor sensor : sensors) {
            sensor.getApplications().clear();
            for (Application application : applications) {
                if (checkIfApplicationHasPermission(application, sensor.getName())) {
                    sensor.addApplication(application);
                }
            }
        }
        return sensors;
    }

    private static List<Application> getApplicationsInformation(Context context) {
        List<Application> applications = new ArrayList<>();
        StringBuilder appNameAndPermissions = new StringBuilder();
        PackageManager pm = context.getPackageManager();
        List<ApplicationInfo> packages = pm.getInstalledApplications(PackageManager.GET_META_DATA);

        for (ApplicationInfo applicationInfo : packages) {
            if (isSystemPackage(applicationInfo))
                continue;
            String name = parseName(applicationInfo.name);
            if (name == null || name.isEmpty())
                continue;
            Application application = new Application(name);
            Log.d("test", "App: " + applicationInfo.name + " Package: " + applicationInfo.packageName);
            try {
                PackageInfo packageInfo = pm.getPackageInfo(applicationInfo.packageName, PackageManager.GET_PERMISSIONS);
                appNameAndPermissions.append(packageInfo.packageName).append("*******:\n");

                //Get Permissions
                String[] requestedPermissions = packageInfo.requestedPermissions;
                if (requestedPermissions != null) {
                    for (String requestedPermission : requestedPermissions) {
                        Log.d("test", requestedPermission);
                        appNameAndPermissions.append(requestedPermission).append("\n");
                        application.addPermission(requestedPermission);
                    }
                }
                appNameAndPermissions.append("\n");
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
            applications.add(application);
        }
        return applications;
    }
}

File: app/src/main/java/com/kent/university/privelt/utils/sensors/SensorHelper.kt
/*
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  * License, v. 2.0. If a copy of the MPL was not distributed with this
 *  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.kent.university.privelt.utils.sensors

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageInfo.REQUESTED_PERMISSION_GRANTED
import android.content.pm.PackageManager
import com.kent.university.privelt.model.Application
import com.kent.university.privelt.model.Sensor
import java.util.*

object SensorHelper {
    fun isSystemPackage(applicationInfo: ApplicationInfo): Boolean {
        return applicationInfo.flags and ApplicationInfo.FLAG_SYSTEM != 0
    }

    private fun checkIfApplicationHasPermission(application: Application, sensor: String): Boolean {
        for (permission in application.getPermissions()) {
            if (permission.toLowerCase(Locale.ROOT).contains(sensor.toLowerCase(Locale.ROOT))) return true
        }
        return false
    }

    fun getNumberOfApplicationInstalled(context: Context): Int {
        var numberOfNonSystemApps = 0

        val appList: List<ApplicationInfo> = context.packageManager.getInstalledApplications(0)
        for (info in appList) {
            if (info.flags and ApplicationInfo.FLAG_SYSTEM == 0) {
                numberOfNonSystemApps++
            }
        }
        return numberOfNonSystemApps
    }

    fun getSensorsInformation(context: Context): List<Sensor> {
        val applications = getApplicationsInformation(context)
        val sensors = listOf(*Sensor.values())
        for (sensor in sensors) {
            sensor.getApplications().clear()
            for (application in applications) {
                if (checkIfApplicationHasPermission(application, sensor.title)) {
                    sensor.addApplication(application)
                }
            }
        }
        return sensors
    }

    fun getIfPermissionWereGranted(pi: PackageInfo, i: Int): Boolean {
        return (pi.requestedPermissionsFlags[i] and REQUESTED_PERMISSION_GRANTED) != 0
    }

    private fun getApplicationsInformation(context: Context): List<Application> {
        val applications: MutableList<Application> = ArrayList()
        val pm = context.packageManager
        val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)
        for (applicationInfo in packages) {
            if (isSystemPackage(applicationInfo)) continue
            val application = Application(applicationInfo.packageName)
            try {
                val packageInfo = pm.getPackageInfo(applicationInfo.packageName, PackageManager.GET_PERMISSIONS)
                val requestedPermissions = packageInfo.requestedPermissions
                if (requestedPermissions != null) {
                    for ((i, requestedPermission) in requestedPermissions.withIndex()) {
                        if (getIfPermissionWereGranted(packageInfo, i))
                            application.addPermission(requestedPermission)
                    }
                }
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }
            applications.add(application)
        }
        return applications
    }
}